docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,335 @@
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+
+        # set policy to present by default, if not specified
+        if "policy" not in attributes:
+            self.log_verbose(f"policy not specified, defaults to present")
+            attributes["policy"] = "present"
+
+        if attributes["policy"] == "present":
+            self.log_debug(f"group '{promiser}' should exist")
+
+            if group is None:
+                self.log_debug(f"group '{promiser}' does not exist")
+                try:
+                    self.log_info(f"creating group '{promiser}'")
+                    Group.create(promiser)
+                    repairs += 1
+                except GroupException as e:
+                    self.log_error(f"failed to create group '{promiser}': {e}")
+                    failed_repairs += 1
+            else:
+                self.log_debug(f"group '{promiser}' does exist")
+
+            group = Group.lookup(promiser)
+            if "gid" in attributes:
+                gid = attributes["gid"]
+
+                if gid != group.gid:
+                    try:
+                        self.log_info(f"setting GID of group '{group.name}' to '{gid}'")
+                        Group.gid_set(group.name, gid)
+                        repairs += 1
+                    except GroupException as e:
+                        self.log_error(f"failed to set GID of group '{group.name}' to '{gid}'")
+                        failed_repairs += 1
+            
+            if "members" in attributes:
+                members = attributes["members"]
+                ret = self._set_members(group, members)
+                failed_repairs += ret[0]
+                repairs += ret[1]
+
+        else:
+            self.log_debug(f"group '{promiser}' should not exist")
+            if group is not None:
+                self.log_debug(f"group '{promiser}' does exist")
+                try:
+                    self.log_info(f"deleting group '{promiser}'")
+                    Group.delete(promiser)
+                    repairs += 1
+                except GroupException as e:
+                    self.log_error(f"failed to delete group '{promiser}': {e}")
+                    failed_repairs += 1
+            else:
+                self.log_debug(f"group {promiser} does not exist")
+
+        self.log_debug(f"{repairs} repairs, {failed_repairs} failed repairs")
+        if failed_repairs > 0:
+            self.log_verbose("promise not kept")
+            return Result.NOT_KEPT
+
+        if repairs > 0:
+            self.log_verbose("promise repaired")
+            return Result.REPAIRED
+
+        self.log_verbose("promise kept")
+        return Result.KEPT
+
+
+    def _set_members(self, group, members):
+        failed_repairs = 0
+        repairs = 0
+
+        for attribute in members:
+            if attribute == "include":
```

Comment:
Please create separate functions for the branches in this `if` and then call the functions based on `if "include" in members"` etc.